Mathlib/Order/LocallyFinite.lean

```diff
@@ -0,0 +1,1380 @@
+    rfl
+
+instance [LocallyFiniteOrderTop α] [LocallyFiniteOrderTop β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] : LocallyFiniteOrderTop (α × β) :=
+  LocallyFiniteOrderTop.ofIci' (α × β) (fun a => Ici a.fst ×ᶠ Ici a.snd) fun a x =>
+    by
+    rw [mem_product, mem_Ici, mem_Ici]
+    rfl
+
+instance [LocallyFiniteOrderBot α] [LocallyFiniteOrderBot β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] : LocallyFiniteOrderBot (α × β) :=
+  LocallyFiniteOrderBot.ofIic' (α × β) (fun a => Iic a.fst ×ᶠ Iic a.snd) fun a x =>
+    by
+    rw [mem_product, mem_Iic, mem_Iic]
+    rfl
+
+theorem Icc_eq [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] (p q : α × β) :
+    Finset.Icc p q = Finset.Icc p.1 q.1 ×ᶠ Finset.Icc p.2 q.2 :=
+  rfl
+#align prod.Icc_eq Prod.Icc_eq
+
+@[simp]
+theorem Icc_mk_mk [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] (a₁ a₂ : α) (b₁ b₂ : β) :
+    Finset.Icc (a₁, b₁) (a₂, b₂) = Finset.Icc a₁ a₂ ×ᶠ Finset.Icc b₁ b₂ :=
+  rfl
+#align prod.Icc_mk_mk Prod.Icc_mk_mk
+
+theorem card_Icc [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] (p q : α × β) :
+    (Finset.Icc p q).card = (Finset.Icc p.1 q.1).card * (Finset.Icc p.2 q.2).card :=
+  Finset.card_product _ _
+#align prod.card_Icc Prod.card_Icc
+
+end Prod
+
+end Preorder
+
+namespace Prod
+
+variable [Lattice α] [Lattice β]
+
+theorem uIcc_eq [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] (p q : α × β) :
+    Finset.uIcc p q = Finset.uIcc p.1 q.1 ×ᶠ Finset.uIcc p.2 q.2 :=
+  rfl
+#align prod.uIcc_eq Prod.uIcc_eq
+
+@[simp]
+theorem uIcc_mk_mk [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] (a₁ a₂ : α) (b₁ b₂ : β) :
+    Finset.uIcc (a₁, b₁) (a₂, b₂) = Finset.uIcc a₁ a₂ ×ᶠ Finset.uIcc b₁ b₂ :=
+  rfl
+#align prod.uIcc_mk_mk Prod.uIcc_mk_mk
+
+theorem card_uIcc [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+    [DecidableRel ((· ≤ ·) : α × β → α × β → Prop)] (p q : α × β) :
+    (Finset.uIcc p q).card = (Finset.uIcc p.1 q.1).card * (Finset.uIcc p.2 q.2).card :=
+  Prod.card_Icc _ _
+#align prod.card_uIcc Prod.card_uIcc
+
+end Prod
+
+/-!
+#### `WithTop`, `WithBot`
+
+Adding a `⊤` to a locally finite `OrderTop` keeps it locally finite.
+Adding a `⊥` to a locally finite `OrderBot` keeps it locally finite.
+-/
+
+
+namespace WithTop
+
+variable (α) [PartialOrder α] [OrderTop α] [LocallyFiniteOrder α]
+
+-- Porting note: removed attribute [local match_pattern] coe
+
+attribute [local simp] Option.mem_iff
+
+private lemma aux (x : α) (p : α → Prop) :
+    (∃ a : α, p a ∧ Option.some a = Option.some x) ↔ p x := by
+  -- Porting note: `simp [Option.some_inj]` has no effect
+  constructor
+  · rintro ⟨x', hx, hx'⟩
+    obtain rfl := Option.some_inj.mp hx'
+    exact hx
+  · exact fun h => ⟨x, h, rfl⟩
```

Comment:
Is this to work around the lack of `.match_1`? The changes here seemed rather verbose, but I haven't tried to simplify.

Reply:
Yeah. I'm not sure what's making this harder in lean4, but I'd prefer not spending too long debugging it.